### PR TITLE
osd, tools, test/osd: implement read balancer

### PR DESCRIPTION
**Pull request by ljflores**
This PR implements two main functions for the read balancer:

1. calc_desired_primary_distribution
   Based on the number of pgs in a pool and the pool's
   replica count, we calculate the ideal number of primary
   pgs that should be assigned to each OSD on that pool in
   order for reads to be balanced.

2. balance_primaries
   This is the overall algorithm used to balance reads (primary
   pgs) in a pool. Based on the first function, we re-distribute
   primaries on the OSDs in a pool so each OSD has the ideal
   number of primaries. This is done without data movement.

This PR also adds the capability to balance reads on a
given osdmap with the osdmaptool. The user has the option
of performing a "dry run" of read balancing OR taking it a
step further and applying the results to a live cluster.

Performing a "dry run" would involve simply running an
osdmaptool command and inspecting the results.

The template for the command is:
`osdmaptool <osd map file> --read <file for command output> --read-pool <pool name>`

An example command a user might run is:
`osdmaptool om --read out.txt --read-pool default.rgw.control`

This commit also adds a `--vstart` flag that allows a user to print ceph
commands in the outfile formatted for a vstart cluster. An example command
a user might run is:
`./bin/osdmaptool om --vstart --read out.txt --read-pool default.rgw.control`
The out.txt file would contain ceph commands prefixed with `./bin/`.
The `--vstart` flag may also be applied to an `--upmap` osdmaptool command.

If the user wants to apply read balancing results from their dry run to a
live cluster, they may either manually apply the ceph commands from the out
file, or run `source <outfile>`.

Lastly, PR adds unit test coverage to the read balancing
feature, including small vs. large osdmap scenarios,
random osdmap scenarios, and scenarios that involve
tweaking primary affinity on an OSD.

Related Trello: https://trello.com/c/is7l6BIb/812-balancer-balance-reads
Related BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1870804
Signed-off-by: Laura Flores <lflores@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by ljflores** [comment]
Note to self: this will be the change needed in the osdmaptool code to print the score when combined with Josh's PR:

```diff
[lflores@folio01 build]$ git diff
diff --git a/src/tools/osdmaptool.cc b/src/tools/osdmaptool.cc
index ccafde9b11a..0dcfb87bee0 100644
--- a/src/tools/osdmaptool.cc
+++ b/src/tools/osdmaptool.cc
@@ -472,10 +472,15 @@ int main(int argc, const char **argv)
     cout << " \n";
     cout << "---------- BEFORE ------------ \n";
     for (auto [osd, pgs] : acting_prims_by_osd) {
-      cout << " osd." << osd << " ~ number of prims: " << pgs.size() << " ~ score: " << "<insert score here>\n";
+      cout << " osd." << osd << " ~ number of prims: " << pgs.size() << "\n";
     }
     cout << " \n";
 
+    OSDMap::read_balance_info_t rb_info;
+    tmp_osd_map.calc_read_balance_score(g_ceph_context, pid, &rb_info);
+    cout << "read_balance_score of " << workload_pool << " : " << rb_info.acting_adj_score << "\n";
+    cout << " \n";
+
     OSDMap::Incremental pending_inc(osdmap.get_epoch()+1);
     int num_changes = osdmap.calc_workload_balancer(g_ceph_context, pid, &pending_inc, tmp_osd_map);
 
@@ -486,9 +491,13 @@ int main(int argc, const char **argv)
     cout << " \n";
     cout << "---------- AFTER ------------ \n";
     for (auto [osd, pgs] : acting_prims_by_osd_2) {
-      cout << " osd." << osd << " ~ number of prims: " << pgs.size() << " ~ score: " << "<insert score here>\n";
+      cout << " osd." << osd << " ~ number of prims: " << pgs.size() << "\n";
     }
     cout << " \n";
+    tmp_osd_map.calc_read_balance_score(g_ceph_context, pid, &rb_info);
+    cout << "read_balance_score of " << workload_pool << " : " << rb_info.acting_adj_score << "\n";
+    cout << " \n";
+
     cout << "num changes: " << num_changes << "\n";
     cout << " \n";

```

**Comment by ljflores** [comment]
[balancer: balance reads](https://trello.com/c/is7l6BIb/812-balancer-balance-reads)

**Comment by ljflores** [comment]
Commits that need to be added after #47555 is merged:

https://github.com/ceph/ceph/commit/05d811670e10ccc916f52305359ec6565694a5d6
https://github.com/ceph/ceph/commit/645e5628c433e8017c9015e91e7e0ecf07e38f3b

**Comment by ljflores** [comment]
I cleaned up the commits. This PR depends on https://github.com/ceph/ceph/pull/47555 to be merged first, which is in testing right now.

After #47555 is merged, I will drop https://github.com/ceph/ceph/pull/48010/commits/3c7ebc2bc485eced28d95f9c7c38b6fc7c458102. It is there for now as a placeholder.

**Review by JoshSalomon — CHANGES_REQUESTED** [review]

**Comment by ljflores** [comment]
Will squash commits after I'm certain make check is passing

File: src/tools/osdmaptool.cc [hunk 1]
```diff
@@ -67,32 +67,48 @@ void usage()
...
+  cout << "   --vstart                prefix upmap and read output with './bin/'" << std::endl;
   exit(1);
 }
 
-void print_inc_upmaps(const OSDMap::Incremental& pending_inc, int fd)
+void print_inc_upmaps(const OSDMap::Incremental& pending_inc, int fd, bool vstart)
 {
   ostringstream ss;
   for (auto& i : pending_inc.old_pg_upmap) {
+    if (vstart)
```
[the remark this answers is not in the document]
Comment (JoshSalomon) [reply src/tools/osdmaptool.cc:81]: You did not implement this one... The idea was to have only one place in which you ask `if (vstart)`

File: src/test/osd/TestOSDMap.cc [hunk 2]
```diff
@@ -2443,12 +2444,261 @@ TEST_F(OSDMapTest, ReadBalanceScore2) {
...
+    osdmap.calc_pg_upmaps(g_ceph_context,
+                          0,
+                          100,
+                          only_pools,
+                          &pending_inc);
+    osdmap.apply_incremental(pending_inc);
+
+    // Get read balance score before balancing
+    OSDMap::read_balance_info_t rb_info;
+    osdmap.calc_read_balance_score(g_ceph_context, my_rep_pool, &rb_info);
```
Comment (JoshSalomon) [inline src/test/osd/TestOSDMap.cc:2479]: I see that you still do not check for the return code of `calc_read_balance_score` in any place.

**Review by JoshSalomon — APPROVED** [review]

**Review by JoshSalomon — APPROVED** [review]

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
Rados suite review:
http://pulpito.front.sepia.ceph.com:80/yuriw-2023-02-22_21:31:50-rados-wip-yuri10-testing-2023-02-22-0848-distro-default-smithi
http://pulpito.front.sepia.ceph.com:80/yuriw-2023-02-23_16:14:52-rados-wip-yuri10-testing-2023-02-22-0848-distro-default-smithi

Failures, unrelated:
    1. https://tracker.ceph.com/issues/57754
    2. https://tracker.ceph.com/issues/58475
    3. https://tracker.ceph.com/issues/58585
    4. https://tracker.ceph.com/issues/58797

Details:
    1. test_envlibrados_for_rocksdb.sh: update-alternatives: error: alternative path /usr/bin/gcc-11 doesn't exist - Ceph - RADOS
    2. test_dashboard_e2e.sh: Conflicting peer dependency: postcss@8.4.21 - Ceph - Mgr - Dashboard
    3. rook: failed to pull kubelet image - Ceph - Orchestrator
    4. scrub/osd-scrub-dump.sh: TEST_recover_unexpected fails from "ERROR: Unexpectedly low amount of scrub reservations seen during test"

Some failures in the runs were directly related to https://github.com/ceph/ceph/pull/49837.

**Comment by batrick** [comment]
Linking to #52547 for posterity.

**Pull request merged**